### ISSUE TITLE
toggle cellular data (mobile data) on phones without root permissions

### ISSUE THREAD

**Issue by toxtox** [renyuneyun/Easer#118]
If you enable "_Use root features_" in "_Settings_" Easer is able to **toggle cellular data**.
But most phones are not rooted.
I guess there is way to **toggle cellular data even on non-rooted phones** because if you install _**LibreTasks**_ https://f-droid.org/packages/libretasks.app/ you can find a permission called **_"toggle cellular data"_** in the "_Settings_" -> "_Privacy Guard_" -> "Advanced" -> "_LibreTask_" section that you can turn on and off. 
Unfortunately _LibreTask_ does not use this permission to toggle cellular data.

(LineageOS14,1, but the "Privacy Guard" for checking this permission exists on other [custom] roms e.g. "MOKEE", too)

**Comment by renyuneyun** [comment]
Hmm, thanks for the info.
(I don't know much about Privacy Guard. Maybe it's time for me to explore a little bit more.)

As you may know, Android doesn't allow user apps to toggle cellular data anymore (after 2.3 I remember), by using normal APIs.
Does the app have to do something special to add that entry (in Privary Guard)? (similarly, is there such an entry for Easer as for now?)
It's quite confusing that Privacy Guard will provide special permissions (overriding Android's later decision) to enable/disable cellular data...

By the way, do you know any documents for using such a special permission (or other ones)?

**Comment by toxtox** [comment]
Yes I know that toggle cellular data is not officially possible on non-routed devices but the permission to toggle cellular data for LibreTask makes me hope that it should work.
As you can see that Easer has no such permission yet.
![screenshot_20180601-211810](https://user-images.githubusercontent.com/15896490/40859890-90c54f90-65e3-11e8-8104-96555f39005d.png)
![screenshot_20180601-211845](https://user-images.githubusercontent.com/15896490/40859893-932eab1e-65e3-11e8-8118-3ade9303e346.png)

I guess you should already know the _"Privacy Guard"_ for setting app permissions because you have written that _"Easer"_ was inspired by _CyanogenMod(LineageOS_)

**Comment by toxtox** [comment]
For testing purposes you could just implement the long list of _LibreTask_ permissions completely in order to check if the "**_toggle cellular data_**" appears in the Privacy Guard settings for your "_Easer_" app.

    android.permission.ACCESS_FINE_LOCATION
    android.permission.ACCESS_COARSE_LOCATION
    android.permission.ACCESS_FINE_LOCATION
    android.permission.ACCESS_LOCATION_EXTRA_COMMANDS
    android.permission.ACCESS_NETWORK_STATE
    android.permission.ACCESS_WIFI_STATE
    android.permission.BATTERY_STATS
    android.permission.BLUETOOTH
    android.permission.BLUETOOTH_ADMIN
    android.permission.CALL_PHONE
    android.permission.CAMERA
    android.permission.CHANGE_COMPONENT_ENABLED_STATE
    android.permission.CHANGE_CONFIGURATION
    android.permission.CHANGE_NETWORK_STATE
    android.permission.CHANGE_WIFI_MULTICAST_STATE
    android.permission.CHANGE_WIFI_STATE
    android.permission.DEVICE_POWER
    android.permission.GET_ACCOUNTS
    android.permission.GLOBAL_SEARCH
    android.permission.INJECT_EVENTS
    android.permission.INTERNET
    android.permission.KILL_BACKGROUND_PROCESSES
    android.permission.MANAGE_ACCOUNTS
    android.permission.MANAGE_APP_TOKENS
    android.permission.MODIFY_AUDIO_SETTINGS
    android.permission.MODIFY_PHONE_STATE
    android.permission.PERSISTENT_ACTIVITY
    android.permission.PROCESS_OUTGOING_CALLS
    android.permission.READ_CALENDAR
    android.permission.READ_CONTACTS
    android.permission.READ_LOGS
    android.permission.READ_OWNER_DATA
    android.permission.READ_PHONE_STATE
    android.permission.READ_SMS
    android.permission.RECEIVE_BOOT_COMPLETED
    android.permission.RECEIVE_MMS
    android.permission.RECEIVE_SMS
    android.permission.SEND_SMS
    android.permission.SET_ACTIVITY_WATCHER
    android.permission.SYSTEM_ALERT_WINDOW
    android.permission.USE_CREDENTIALS
    android.permission.VIBRATE
    android.permission.WRITE_SETTINGS
    android.permission.WRITE_SMS

**Comment by renyuneyun** [comment]
Hmm, maybe I have a guess. I'll try to confirm it later, but it seems to me that there is no difference (to a standard Android allowance).

Yes, I had a period (for my previous phone) using CyanogenMod, but that's a few years ago (I remember at that time CM was 12 or 13 at most). I wasn't aware Privacy Guard is provided by CM but not Android, until you opened this issue :p

**Comment by toxtox** [comment]
Have you already made some tests to get the "**_toggle cellular data_**" permission?

By the way I have found two other apps that have this permission:
https://f-droid.org/packages/org.dnaq.dialer2/
![screenshot_20180609-165943](https://user-images.githubusercontent.com/15896490/41193057-ca07f8d2-6c07-11e8-854f-0b59fee6b6df.png)
according to "f-droid.org" the "dialer2" app has only very few permission:

    android.permission.CALL_PHONE
    android.permission.CALL_PRIVILEGED
    android.permission.READ_CONTACTS
    android.permission.MODIFY_PHONE_STATE
    android.permission.WRITE_CONTACTS
    android.permission.VIBRATE
    android.permission.READ_CALL_LOG
    android.permission.WRITE_CALL_LOG


and the standard phone app of LineageOS 14.1 has the "**_toggle cellular data_**", too
![screenshot_20180609-170251](https://user-images.githubusercontent.com/15896490/41193071-eee75da0-6c07-11e8-8069-0e7f38fc6d5d.png)

**Comment by renyuneyun** [comment]
Hi, thanks for digging further.
I now see how they got this "permission" listed in the list, but my conclusion indicates that they can't actually change it...

They use the [`MODIFY_PHONE_STATE`](https://developer.android.com/reference/android/Manifest.permission#MODIFY_PHONE_STATE) permission, and Android document says this permission isn't meant for third-party apps to use. This means third-party apps (e.g. Easer) can't get this permission.
The only way is to make Easer a system app, but this requires ROM making or root permission (as always...)...
